src/persist-client/src/internal/gc.rs

```diff
@@ -181,58 +216,113 @@ impl GarbageCollector {
+                        // previous versions of state.
+                        deleteable_batch_blobs.insert(key.to_owned());
+                    }
                 });
-            } else if state_version.seqno == req.new_seqno_since {
-                Self::for_all_keys(&state_version, |key| {
-                    // It's okay (expected) if the key doesn't exist in
-                    // deleteable_blobs, it may have been added in this version
-                    // of state.
-                    let _ = deleteable_blobs.remove(key);
+            } else if diff.seqno == req.new_seqno_since {
+                fetcher.apply_diffs(&mut state, std::iter::once(&diff));
+                assert_eq!(state.seqno, diff.seqno);
+                state.collections.trace.map_batches(|b| {
+                    for key in b.keys.iter() {
+                        // It's okay (expected) if the key doesn't exist in
+                        // deleteable_blobs, it may have been added in this version
+                        // of state.
+                        let _ = deleteable_batch_blobs.remove(key);
+                    }
                 });
+                // We only need to detect deletable rollups in the last iter
+                // through the live_diffs loop because they accumulate in state.
+                for (seqno, key) in state.collections.rollups.iter() {
+                    // SUBTLE: This is intentionally comparing vs
+                    // old_seqno_since, not new_seqno_since. We could collect
+                    // all rollups less than new_seqno_since, and then remove
+                    // them from state _after the truncate_, but the state
+                    // change to add the new rollups has to be before the
+                    // truncate and we don't want to create 2 state changes per
+                    // call into gc. In a way, we
+                    if seqno < &req.old_seqno_since {
+                        deleteable_rollup_blobs.push((*seqno, key.clone()));
+                    } else {
+                        // We iterate in order, may as well short circuit the
+                        // rollup loop.
+                        break;
+                    }
+                }
             } else {
                 // Sanity check the loop logic.
-                assert!(state_version.seqno > req.new_seqno_since);
+                assert!(diff.seqno > req.new_seqno_since);
                 break;
             }
         }
 
+        // We maintain the invariant that there is always a rollup corresponding
+        // to the seqno of the first live version in consensus. Write a new
+        // rollup at exactly req.new_seqno_since so we're free to truncate
+        // anything before it.
+        //
+        // It's unfortunate to tie the rollups schedule to gc so tightly, but
+        // the above invariant makes our leaked blob logic easier.
```

Comment:
if GC is held back for a long time, would we ever want to force a state rollup to be written? otherwise new handles may need to run arbitrarily long scans through consensus to reconstruct the latest state if the seqno since is held back

Reply:
also I think I'm missing a key insight into how this invariant makes leaked blob detection simpler 😓

Reply:
> if GC is held back for a long time, would we ever want to force a state rollup to be written? otherwise new handles may need to run arbitrarily long scans through consensus to reconstruct the latest state if the seqno since is held back

yeah, but I think we have to cut scope at some point and get this merged. adding a separate schedule for rollups is inevitable, but it's a decent chunk of work and I'm not sure it's worth missing this next release over

> also I think I'm missing a key insight into how this invariant makes leaked blob detection simpler 😓

explained in the big StateVersions rustdoc (and a pointer to it put here). summary: any rollup < the earliest seqno is consensus has been leaked

Reply:
> I think we can just put this as another Maintenance task that triggers when there's a big delta between latest_rollup and seqno
super straightforward. done!